# OLA RIDES ANALYSIS 
## PROBLEM STATEMENT
##### OLA, a leading ride-hailing service, generates vast amounts of data related to ride bookings, driver availability, fare calculations, and customer preferences. However, deriving actionable insights from this data remains a challenge. To enhance operational efficiency, improve customer satisfaction, and optimize business strategies, this project focuses on analyzing OLA’s ride-sharing data. By leveraging data analytics, visualization techniques, and interactive applications, the goal is to extract meaningful insights that can drive data-informed decisions. The project will involve cleaning and processing raw ride data, performing exploratory data analysis (EDA), developing a dynamic Power BI dashboard, and creating a Streamlit-based web application to present key findings in an interactive and user-friendly manner.

## PROBLEM SOLUTION 
### Data Understanding & Exploration
##### * Load and examine the dataset structure.
##### * Identify key variables like ride status, payment method, and ratings.
##### * Perform initial exploratory data analysis (EDA).
### Data Cleaning & Preprocessing
##### * Handle missing or inconsistent values.
##### * Convert data types and standardize formats.
##### * Create derived features if necessary for better insights(created new column like hour of the day and day of the week)
### SQL Query Development
##### * Write queries to extract insights (e.g., ride trends, cancellations, ratings).
##### * Optimize queries for performance and accuracy.
##### * Validate results against the dataset.
### Power BI Dashboard Creation
##### * Design interactive visualizations for ride trends, revenue, and cancellations.
##### * Integrate KPIs and metrics for business insights.
### Streamlit Application Development
##### * Create a user-friendly UI to display SQL query results.
##### * Implement interactive filters and search options.
### Project Documentation & Deployment
##### * Document insights, queries, and dashboard explanations.
##### * Ensure the Streamlit app is deployed and accessible.
##### * Present findings with business-oriented storytelling.



##### Github link:

### Let's begin

Importing the neccesary libraries

In [1]:
# Loading Libraries
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns


Load the dataset for EDA

In [2]:
# read the file
df = pd.read_csv('C:\\Users\\anous\\OneDrive\\PROJECTS\\OLA RIDES ANALYSIS\\OLA_DataSet.xlsx - July.csv')
df.head()

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00   3:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00   9:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0   

                     Canceled_Rides_by_Customer      Canceled_Rides_by_Driver  \
0                                           NaN  Personal & Car related issue   
1                                           NaN                           NaN   
2                                           NaN                           NaN   
3  Driver is not moving towards pickup location                           NaN   
4                                           NaN                           NaN   

  Incomplete_Rides Incomplete_Rides_Reason  Booking_Value Payment_Method  \
0              NaN                     NaN            444            NaN   
1               No                     NaN            158           Cash   
2               No                     NaN            386            UPI   
3              NaN                     NaN            384            NaN   
4               No                     NaN            822    Credit Card   

   Ride_Distance  Driver_Ratings  Customer_Rating  \
0              0             NaN              NaN   
1             13             4.1              4.0   
2             40             4.2              4.8   
3              0             NaN              NaN   
4             45             4.0              3.0   

                                      Vehicle Images  
0  https://cdn-icons-png.flaticon.com/128/14183/1...  
1  https://cdn-icons-png.flaticon.com/128/9983/99...  
2  https://cdn-icons-png.flaticon.com/128/9983/99...  
3  https://cdn-icons-png.flaticon.com/128/6839/68...  
4  https://cdn-icons-png.flaticon.com/128/3202/32...

Checking the null values, missing values, duplicates and datatype for more clear understanding of dataset and remove unneccessary values.

In [3]:
df.isnull().sum()

Date                              0
Time                              0
Booking_ID                        0
Booking_Status                    0
Customer_ID                       0
Vehicle_Type                      0
Pickup_Location                   0
Drop_Location                     0
V_TAT                         39057
C_TAT                         39057
Canceled_Rides_by_Customer    92525
Canceled_Rides_by_Driver      84590
Incomplete_Rides              39057
Incomplete_Rides_Reason       99098
Booking_Value                     0
Payment_Method                39057
Ride_Distance                     0
Driver_Ratings                39057
Customer_Rating               39057
Vehicle Images                    0
dtype: int64

In [4]:
# checking for duplicate values
df.duplicated().sum()

np.int64(0)

As it shown in the following datatype the date and time are objet we change this to correct datatypes in datetime format

In [5]:
# check the datatypes
df.dtypes

Date                           object
Time                           object
Booking_ID                     object
Booking_Status                 object
Customer_ID                    object
Vehicle_Type                   object
Pickup_Location                object
Drop_Location                  object
V_TAT                         float64
C_TAT                         float64
Canceled_Rides_by_Customer     object
Canceled_Rides_by_Driver       object
Incomplete_Rides               object
Incomplete_Rides_Reason        object
Booking_Value                   int64
Payment_Method                 object
Ride_Distance                   int64
Driver_Ratings                float64
Customer_Rating               float64
Vehicle Images                 object
dtype: object

Convert time and date column, and extract new features like HOUR OF THE DAY and DAY OF THE WEEK

In [6]:
# Convert 'Date' and 'Time' columns; extract new features: day of week and hour of day
df['Date'] = pd.to_datetime(df['Date'])
df['DayOfWeek'] = df['Date'].dt.day_name()
df['Time'] = pd.to_datetime(df['Time'], errors='coerce')
df['Hour'] = df['Time'].dt.hour.fillna(0).astype(int)

C:\Users\anous\AppData\Local\Temp\ipykernel_17672\2304098489.py:4: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['Time'] = pd.to_datetime(df['Time'], errors='coerce')


In [7]:
# Replace missing values with constant values
# For numeric columns, fill with 0; for object/categorical columns, fill with 'Unknown'
for col in df.columns:
    if df[col].dtype in ['float64', 'int64']:
        df[col].fillna(0, inplace=True)
    else:
        df[col].fillna('Unknown', inplace=True)

df.isnull().sum()

C:\Users\anous\AppData\Local\Temp\ipykernel_17672\3982819648.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna('Unknown', inplace=True)
C:\Users\anous\AppData\Local\Temp\ipykernel_17672\3982819648.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, whe

Date                          0
Time                          0
Booking_ID                    0
Booking_Status                0
Customer_ID                   0
Vehicle_Type                  0
Pickup_Location               0
Drop_Location                 0
V_TAT                         0
C_TAT                         0
Canceled_Rides_by_Customer    0
Canceled_Rides_by_Driver      0
Incomplete_Rides              0
Incomplete_Rides_Reason       0
Booking_Value                 0
Payment_Method                0
Ride_Distance                 0
Driver_Ratings                0
Customer_Rating               0
Vehicle Images                0
DayOfWeek                     0
Hour                          0
dtype: int64

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 22 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Date                        103024 non-null  datetime64[ns]
 1   Time                        103024 non-null  datetime64[ns]
 2   Booking_ID                  103024 non-null  object        
 3   Booking_Status              103024 non-null  object        
 4   Customer_ID                 103024 non-null  object        
 5   Vehicle_Type                103024 non-null  object        
 6   Pickup_Location             103024 non-null  object        
 7   Drop_Location               103024 non-null  object        
 8   V_TAT                       103024 non-null  float64       
 9   C_TAT                       103024 non-null  float64       
 10  Canceled_Rides_by_Customer  103024 non-null  object        
 11  Canceled_Rides_by_Driver    103024 non-

Here we performed the statistical summary

In [9]:
df.describe()

Date                           Time  \
count                         103024                         103024   
mean   2024-07-16 11:31:38.879678720  2025-09-28 11:58:37.453991424   
min              2024-07-01 00:00:00            2025-09-28 00:00:00   
25%              2024-07-08 18:41:00            2025-09-28 06:01:00   
50%              2024-07-16 11:23:00            2025-09-28 11:58:00   
75%              2024-07-24 05:18:00            2025-09-28 17:57:00   
max              2024-07-31 23:58:00            2025-09-28 23:59:00   
std                              NaN                            NaN   

               V_TAT          C_TAT  Booking_Value  Ride_Distance  \
count  103024.000000  103024.000000  103024.000000  103024.000000   
mean      106.096502      52.697381     548.751883      14.189927   
min         0.000000       0.000000     100.000000       0.000000   
25%         0.000000       0.000000     242.000000       0.000000   
50%        84.000000      45.000000     386.000000       8.000000   
75%       196.000000      95.000000     621.000000      26.000000   
max       308.000000     145.000000    2999.000000      49.000000   
std       104.532203      50.005090     536.541221      15.776270   

       Driver_Ratings  Customer_Rating           Hour  
count   103024.000000    103024.000000  103024.000000  
mean         2.481997         2.482529      11.485528  
min          0.000000         0.000000       0.000000  
25%          0.000000         0.000000       6.000000  
50%          3.400000         3.400000      11.000000  
75%          4.200000         4.200000      17.000000  
max          5.000000         5.000000      23.000000  
std          1.991983         1.992770       6.912131

## The final dataset is ready for further analysis

In [10]:
df.head()

Date                Time     Booking_ID  \
0 2024-07-26 14:00:00 2025-09-28 14:00:00  CNR7153255142   
1 2024-07-25 22:20:00 2025-09-28 22:20:00  CNR2940424040   
2 2024-07-30 19:59:00 2025-09-28 19:59:00  CNR2982357879   
3 2024-07-22 03:15:00 2025-09-28 03:15:00  CNR2395710036   
4 2024-07-02 09:02:00 2025-09-28 09:02:00  CNR1797421769   

         Booking_Status Customer_ID Vehicle_Type Pickup_Location  \
0    Canceled by Driver   CID713523  Prime Sedan     Tumkur Road   
1               Success   CID225428         Bike     Magadi Road   
2               Success   CID270156    Prime SUV   Sahakar Nagar   
3  Canceled by Customer   CID581320        eBike      HSR Layout   
4               Success   CID939555         Mini     Rajajinagar   

  Drop_Location  V_TAT  C_TAT  ... Incomplete_Rides Incomplete_Rides_Reason  \
0      RT Nagar    0.0    0.0  ...          Unknown                 Unknown   
1       Varthur  203.0   30.0  ...               No                 Unknown   
2       Varthur  238.0  130.0  ...               No                 Unknown   
3   Vijayanagar    0.0    0.0  ...          Unknown                 Unknown   
4   Chamarajpet  252.0   80.0  ...               No                 Unknown   

  Booking_Value Payment_Method  Ride_Distance Driver_Ratings  Customer_Rating  \
0           444        Unknown              0            0.0              0.0   
1           158           Cash             13            4.1              4.0   
2           386            UPI             40            4.2              4.8   
3           384        Unknown              0            0.0              0.0   
4           822    Credit Card             45            4.0              3.0   

                                      Vehicle Images  DayOfWeek Hour  
0  https://cdn-icons-png.flaticon.com/128/14183/1...     Friday   14  
1  https://cdn-icons-png.flaticon.com/128/9983/99...   Thursday   22  
2  https://cdn-icons-png.flaticon.com/128/9983/99...    Tuesday   19  
3  https://cdn-icons-png.flaticon.com/128/6839/68...     Monday    3  
4  https://cdn-icons-png.flaticon.com/128/3202/32...    Tuesday    9  

[5 rows x 22 columns]

In [11]:
# Download the cleaned dataset as CSV in Jupyter Notebook
#df_cleaned.to_csv('cleaned_ola_rides.csv', index=False)
#print('Dataset saved as cleaned_ola_rides.csv')

 Now, we have cleaned "OLA RIDES DATASET- JULY", and named it "cleaned_ola_rides.csv" so will be working on this new dataset. 

 Working on "cleaned_ola_rides.csv".

### SQL Query Development
 From here we begin our SQL queries


In [12]:
# Connect to SQLite Database and import CSV data
import sqlite3
import pandas as pd

# Load the CSV data
df = pd.read_csv('C:\\Users\\anous\\OneDrive\\PROJECTS\\OLA RIDES ANALYSIS\\cleaned_ola_rides.csv')

# Connect to SQLite (creates file if it doesn't exist)
conn = sqlite3.connect('ola_rides.db')

# Write DataFrame to SQL table 'rides'
df.to_sql('rides', conn, if_exists='replace', index=False)
print('SQLite database created and data imported!')

SQLite database created and data imported!


In [13]:
# Show all columns in the rides table
print(df.columns.tolist())

['Date', 'Time', 'Booking_ID', 'Booking_Status', 'Customer_ID', 'Vehicle_Type', 'Pickup_Location', 'Drop_Location', 'V_TAT', 'C_TAT', 'Canceled_Rides_by_Customer', 'Canceled_Rides_by_Driver', 'Incomplete_Rides', 'Incomplete_Rides_Reason', 'Booking_Value', 'Payment_Method', 'Ride_Distance', 'Driver_Ratings', 'Customer_Rating', 'Vehicle Images', 'DayOfWeek', 'Hour']


In [14]:
# Example: Run SQL queries from notebook and display results as DataFrame
def run_query(query):
    return pd.read_sql_query(query, conn)



## 1. Retrieve all successful bookings

In [15]:
# 1. Retrieve all successful bookings:
print("1. Retrieve all successful bookings:")
query1 = "SELECT COUNT(*) AS successful_bookings FROM rides WHERE Booking_Status = 'Success';"
df_successful = run_query(query1)
display(df_successful.head())

1. Retrieve all successful bookings:


successful_bookings
0                63967

##  2. Find the average ride distance for each vehicle type:


In [16]:

# 2. Find the average ride distance for each vehicle type:
print("2. Find the average ride distance for each vehicle type:")
query2 = "SELECT Vehicle_Type, AVG(Ride_Distance) AS avg_distance FROM rides GROUP BY Vehicle_Type;"
df_avg_distance = run_query(query2)
display(df_avg_distance)


2. Find the average ride distance for each vehicle type:


Vehicle_Type  avg_distance
0         Auto      6.238089
1         Bike     15.533079
2         Mini     15.510102
3   Prime Plus     15.447474
4    Prime SUV     15.274514
5  Prime Sedan     15.764939
6        eBike     15.580589

## 3. List the top 5 customers who booked the highest number of rides:


In [17]:

# 3. List the top 5 customers who booked the highest number of rides:
print("3. List the top 5 customers who booked the highest number of rides:")
query3 = "SELECT Customer_ID, COUNT(*) AS ride_count FROM rides GROUP BY Customer_ID ORDER BY ride_count DESC LIMIT 5;"
df_top_customers = run_query(query3)
display(df_top_customers)

3. List the top 5 customers who booked the highest number of rides:


Customer_ID  ride_count
0   CID954071           5
1   CID980727           4
2   CID969725           4
3   CID966929           4
4   CID952434           4

## 4.  For the top 5 customers with the highest number of rides, show what vehicle types they used


In [18]:
# 4.  For the top 5 customers with the highest number of rides, show what vehicle types they used
print("Vehicle types used by the top 5 customers with the highest number of rides:")
query4 = '''
WITH top_customers AS (
    SELECT Customer_ID
    FROM rides
    GROUP BY Customer_ID
    ORDER BY COUNT(*) DESC
    LIMIT 5
)
SELECT r.Customer_ID, r.Vehicle_Type, COUNT(*) AS ride_count
FROM rides r
JOIN top_customers t ON r.Customer_ID = t.Customer_ID
GROUP BY r.Customer_ID, r.Vehicle_Type
ORDER BY r.Customer_ID, ride_count DESC;
'''
df_top5_vehicle_types = run_query(query4)
display(df_top5_vehicle_types)

Vehicle types used by the top 5 customers with the highest number of rides:


Customer_ID Vehicle_Type  ride_count
0    CID952434         Bike           4
1    CID954071  Prime Sedan           2
2    CID954071         Mini           2
3    CID954071    Prime SUV           1
4    CID966929         Mini           3
5    CID966929    Prime SUV           1
6    CID969725         Mini           2
7    CID969725    Prime SUV           1
8    CID969725   Prime Plus           1
9    CID980727    Prime SUV           2
10   CID980727         Bike           2

## 5. Show the count of each booking status


In [19]:
# 5. Show the count of each booking status
print("Count of each booking status:")
query5 = "SELECT Booking_Status, COUNT(*) AS status_count FROM rides GROUP BY Booking_Status;"
df_booking_status_counts = run_query(query5)
display(df_booking_status_counts)

Count of each booking status:


Booking_Status  status_count
0  Canceled by Customer         10499
1    Canceled by Driver         18434
2      Driver Not Found         10124
3               Success         63967

## 6. Find the average driver ratings for Vehicle type bookings:


In [20]:
# 6. Find the average driver ratings for Vehicle type bookings:
print("6. Find the average and minimum driver ratings for Vehicle type bookings:")
query6 = "SELECT AVG(Driver_Ratings) AS avg_rating, MIN(Driver_Ratings) AS min_rating FROM rides;"
df_vehicle_ratings = run_query(query6)
display(df_vehicle_ratings)


6. Find the average and minimum driver ratings for Vehicle type bookings:


avg_rating  min_rating
0    2.481997         0.0

## 7. Show the average customer rating for each vehicle type, sorted from highest to lowest


In [21]:
# 7. Show the average customer rating for each vehicle type, sorted from highest to lowest
print("Average customer rating for each vehicle type (highest to lowest):")
query7 = "SELECT Vehicle_Type, AVG(Customer_Rating) AS avg_customer_rating FROM rides GROUP BY Vehicle_Type ORDER BY avg_customer_rating DESC;"
df_avg_rating_per_vehicle = run_query(query7)
display(df_avg_rating_per_vehicle)

Average customer rating for each vehicle type (highest to lowest):


Vehicle_Type  avg_customer_rating
0  Prime Sedan             2.522747
1         Bike             2.487757
2         Auto             2.484385
3         Mini             2.482374
4   Prime Plus             2.474074
5        eBike             2.470876
6    Prime SUV             2.455026

## 8. Calculate the count of booking value of rides completed successfully


In [22]:

# 8. Calculate the count of booking value of rides completed successfully
print("Total booking value of rides completed successfully:")
query8 = "SELECT count(Booking_Value) AS total_successful_booking_value_completed_successfully FROM rides WHERE Booking_Status = 'Success' ;"
df_total_booking_value_completed_successfully = run_query(query8)
display(df_total_booking_value_completed_successfully)

Total booking value of rides completed successfully:


total_successful_booking_value_completed_successfully
0                                              63967

## 9. Calculate the total booking value of rides Cancelled by customers


In [23]:
# 9. Calculate the total booking value of rides Cancelled by customers
print("Total booking value of rides Cancelled by customers:")
query9 = "SELECT count(Booking_Value) AS total_cancelled_booking_value_cancelled_by_customers FROM rides WHERE Booking_Status = 'Canceled by Customer' ;"
df_total_booking_value_cancelled_by_customers = run_query(query9)
display(df_total_booking_value_cancelled_by_customers)

Total booking value of rides Cancelled by customers:


total_cancelled_booking_value_cancelled_by_customers
0                                              10499


## 10. Calculate the total booking value of rides Driver Not Found


In [24]:
# 10. Calculate the total booking value of rides Driver Not Found
print("Total booking value of rides Driver Not Found:")
query10 = "SELECT count(Booking_Value) AS total_cancelled_booking_value_driver_not_found FROM rides WHERE Booking_Status = 'Driver Not Found' ;"
df_total_booking_value_driver_not_found = run_query(query10)
display(df_total_booking_value_driver_not_found)

Total booking value of rides Driver Not Found:


total_cancelled_booking_value_driver_not_found
0                                           10124

## 11. Calculate the total booking value of rides Cancelled by drivers


In [25]:
# 11. Calculate the total booking value of rides Cancelled by drivers
print("Total booking value of rides Cancelled by drivers:")
query11 = "SELECT count(Booking_Value) AS total_cancelled_booking_value_cancelled_by_drivers FROM rides WHERE Booking_Status = 'Canceled by Driver' ;"
df_total_booking_value_cancelled_by_drivers = run_query(query11)
display(df_total_booking_value_cancelled_by_drivers)

Total booking value of rides Cancelled by drivers:


total_cancelled_booking_value_cancelled_by_drivers
0                                              18434

## 12. Show the total count of each payment method (excluding 'Unknown')


In [26]:
# 12. Show the total count of each payment method (excluding 'Unknown')
print("Total count of each payment method (excluding 'Unknown'):")
query12 = "SELECT Payment_Method, COUNT(*) AS method_count FROM rides WHERE Payment_Method != 'Unknown' GROUP BY Payment_Method;"
df_payment_method_counts = run_query(query12)
display(df_payment_method_counts)

Total count of each payment method (excluding 'Unknown'):


Payment_Method  method_count
0           Cash         35022
1    Credit Card          2435
2     Debit Card           629
3            UPI         25881

## 13. what is the total number of rides for each day of the week? (descending order)


In [27]:
# 13. what is the total number of rides for each day of the week? (descending order)
print("Total number of rides for each day of the week:")
day_map = {'0': 'Sunday', '1': 'Monday', '2': 'Tuesday', '3': 'Wednesday', '4': 'Thursday', '5': 'Friday', '6': 'Saturday'}
query13 = "SELECT strftime('%w', Date) AS DayOfWeekNum, COUNT(*) AS total_rides FROM rides GROUP BY DayOfWeekNum ORDER BY total_rides DESC;"
df_total_rides_per_day = run_query(query13)
df_total_rides_per_day['DayOfWeek'] = df_total_rides_per_day['DayOfWeekNum'].map(day_map)
df_total_rides_per_day = df_total_rides_per_day[['DayOfWeek', 'total_rides']]
display(df_total_rides_per_day)

Total number of rides for each day of the week:


DayOfWeek  total_rides
0        NaN        42976
1     Monday         9843
2    Tuesday         9786
3  Wednesday         9519
4     Friday         7865
5   Thursday         7747
6   Saturday         7682
7     Sunday         7606

## 14. Total highest booking done every hour of the day


In [28]:
# 14. Total highest booking done every hour of the day
print("Total highest booking done every hour of the day:")
query14 = "SELECT Hour AS booking_hour, COUNT(*) AS total_bookings FROM rides GROUP BY booking_hour;"
df_total_bookings_per_hour = run_query(query14)
display(df_total_bookings_per_hour)


Total highest booking done every hour of the day:


booking_hour  total_bookings
0              0            4318
1              1            4329
2              2            4305
3              3            4201
4              4            4263
5              5            4264
6              6            4271
7              7            4304
8              8            4374
9              9            4347
10            10            4334
11            11            4272
12            12            4408
13            13            4285
14            14            4232
15            15            4376
16            16            4261
17            17            4353
18            18            4176
19            19            4271
20            20            4228
21            21            4343
22            22            4283
23            23            4226

## 📊 Data Insights, Challenges & Recommendations
### Key Data Insights

#### Ride Trends: Out of 103,024 rides, only 62.1% were successful, while 28.1% were canceled (10,499 by customers, 18,434 by drivers).

#### Vehicle Usage: Prime Sedan, eBike, and Bike together account for the highest total ride distance, highlighting both premium ride demand and last-mile connectivity.

#### Ratings: Both average customer and driver ratings are 2.48/5, reflecting a clear satisfaction gap and potential service quality issues.

### Revenue Insights:

#### “Unknown” payment method accounts for 37.9% of revenue → strong indication of data quality gaps.

#### Among valid payments, Cash (34.1%) and UPI (25.1%) dominate transactions.

#### Customer Value: Top 5 customers alone generated over ₹32K in bookings, showing opportunities for loyalty and rewards programs.


### Challenges Identified
 
#### High Cancellation Rate (28%) → leading to lost revenue and poor customer experience.

#### Low Ratings (2.48/5) → indicates dissatisfaction, mismatch in expectations, or rating misuse.

#### Data Quality Issues → large portion of “Unknown” payment methods and cancellation reasons.

#### Revenue Concentration → heavy reliance on cash/UPI, limited digital diversity.

#### Customer Retention Risk → revenue is skewed, with a few top customers contributing significantly.

#### Operational Imbalance → more cancellations by drivers than customers, suggesting supply/demand mismatches.

## Recommendations

### Reduce Cancellations:

#### Improve driver allocation algorithms to reduce long pickup times.

#### Introduce incentives/penalties for both drivers and customers to discourage unnecessary cancellations.

#### Improve Ratings & Experience:

#### Launch training/feedback programs for drivers.

#### Use ride feedback forms to capture detailed service issues.

#### Reward consistently high-rated drivers to motivate better service.

#### Fix Data Quality Issues:

#### Standardize payment recording systems to eliminate “Unknown” entries.

#### Redefine and enforce structured cancellation reasons.

#### Revenue Diversification:

#### Promote digital payments with offers (UPI, cards, wallets).

#### Reduce dependency on cash by introducing cashless incentives.

#### Customer Loyalty Strategy:

#### Identify top spenders and roll out loyalty programs (discounts, premium access).

#### Use referral bonuses to expand high-value customer base.